### PR TITLE
Feature/student audit it

### PR DESCRIPTION
**Pull request by chdietz**
Test audit_student
include inferred school

### CODE REVIEW

File: exam-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/repository/impl/StudentRepositoryIT.java [hunk 1]
```diff
@@ -242,4 +247,137 @@ public void itShouldHandleNullableDatesComparison() throws ParseException {
...
+        // Save original student to compare
+        Map<String, Object> secondStudentToUpdate = namedParameterTemplate.queryForMap(
+                "SELECT * FROM student WHERE id = :id",
+                new MapSqlParameterSource().addValue("id", originalStudentId));
+
+        // Test with exam existing for student to update inferred school
+        jdbcTemplate.update("INSERT INTO asmt (id, natural_id, grade_id, type_id, subject_id, school_year, import_id, update_import_id, name, label) VALUES " +
+                "(-20, 'ASMT20', 11, 1, 1, 2017, -99, -99, 'name', 'label');");
+        jdbcTemplate.update("INSERT INTO exam (id, type_id, school_year, asmt_id, oppId, completeness_id, administration_condition_id, completed_at, import_id, update_import_id, session_id, grade_id, student_id, school_id, iep, lep, section504, economic_disadvantage) VALUES " +
+                "(-399, 1, 2017, -20, '1021', 2, 1, '2017-05-12', -99, -99, 'session', 11, " + originalStudentId + " , -27, 0, 0, 0, 0);");
```
Comment (chdietz) [inline exam-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/repository/impl/StudentRepositoryIT.java]: Sets up the student upsert to update inferred school and the date that goes with it.

File: exam-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/repository/impl/StudentRepositoryIT.java [hunk 2]
```diff
@@ -242,4 +247,137 @@ public void itShouldHandleNullableDatesComparison() throws ParseException {
...
+                        .lastOrSurname("LastNameChanged")
+                        .birthday(LocalDate.parse("2006-02-03"))
+                        .build(), -96);
+
+        Map<String, Object> thirdAuditValues = namedParameterTemplate.queryForMap(
+                "SELECT * FROM audit_student WHERE student_id = :student_id ORDER BY audited DESC LIMIT 1",
+                new MapSqlParameterSource().addValue("student_id", originalStudentId));
+
+        // At least on comparison between audit and actual must have no null values in actual to assure all columns are audited.
+        thirdStudentToUpdate.forEach((k, v) -> assertThat(v).isNotNull());
```
Comment (chdietz) [inline exam-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/repository/impl/StudentRepositoryIT.java]: The actual to compare to audit has no null columns so we know every column is in the audit table and in the trigger sql.

File: exam-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/repository/impl/StudentRepositoryIT.java [hunk 3]
```diff
@@ -242,4 +247,137 @@ public void itShouldHandleNullableDatesComparison() throws ParseException {
...
+
+        assertThat(originalStudentId).isEqualTo(fourthStudentId);
+        assertThat(countRowsInTableWhere(jdbcTemplate, "audit_student", " student_id = " + originalStudentId)).isEqualTo(3);
+        assertThat(countRowsInTableWhere(jdbcTemplate, "student", " id = " + originalStudentId)).isEqualTo(1);
+        assertThat(thirdAuditValues.remove("id")).isNotNull();
+        assertThat(thirdAuditValues.remove("action")).isEqualTo("update");
+        assertThat(thirdAuditValues.remove("audited")).isNotNull();
+        assertThat(thirdAuditValues.remove("database_user")).isNotNull();
+        assertThat(thirdAuditValues.remove("student_id")).isEqualTo(thirdStudentToUpdate.remove("id"));
+        assertThat(thirdStudentToUpdate).isEqualTo(thirdAuditValues);
```
Comment (chdietz) [inline exam-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/repository/impl/StudentRepositoryIT.java]: Compare every column except extra columns in audit table and the id checked independently

File: exam-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/repository/impl/StudentRepositoryIT.java [hunk 4]
```diff
@@ -242,4 +247,137 @@ public void itShouldHandleNullableDatesComparison() throws ParseException {
...
+
+        // *** Second update to student ***
+        // Save original student to compare
+        Map<String, Object> secondStudentToUpdate = namedParameterTemplate.queryForMap(
+                "SELECT * FROM student WHERE id = :id",
+                new MapSqlParameterSource().addValue("id", originalStudentId));
+
+        // Test with exam existing for student to update inferred school
+        jdbcTemplate.update("INSERT INTO asmt (id, natural_id, grade_id, type_id, subject_id, school_year, import_id, update_import_id, name, label) VALUES " +
+                "(-20, 'ASMT20', 11, 1, 1, 2017, -99, -99, 'name', 'label');");
```
Comment (agorina) [inline exam-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/repository/impl/StudentRepositoryIT.java]: Ditto.

File: exam-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/repository/impl/StudentRepositoryIT.java [hunk 5]
```diff
@@ -242,4 +247,137 @@ public void itShouldHandleNullableDatesComparison() throws ParseException {
...
+                    "( -98, 0, 1, 'type', 'digest')," +
+                    "( -97, 0, 1, 'type', 'digest')," +
+                    "( -96, 0, 1, 'type', 'digest')," +
+                    "( -95, 0, 1, 'type', 'digest')",
+            "INSERT INTO district (id, name, natural_id) VALUES (-22, 'Sample District 1', 'DistrictNaturalId');",
+            "INSERT INTO school (id, district_id, name, natural_id, import_id, update_import_id) VALUES (-27, -22, 'Sample School 1', 'SchoolNaturalId', -99, -99);",
+    })
+    public void itShouldAuditStudent() throws ParseException {
+        jdbcTemplate.update("UPDATE setting s SET s.value = 'TRUE' WHERE s.name = 'AUDIT_TRIGGER_ENABLE';");
+        assertThat(countRowsInTableWhere(jdbcTemplate, "setting s", " s.name = 'AUDIT_TRIGGER_ENABLE' and s.value = 'TRUE'")).isEqualTo(1);
```
[the remark this answers is not in the document]
Comment (chdietz) [reply exam-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/repository/impl/StudentRepositoryIT.java]: Maybe not, every test 'tests' auditing will look at this in next test I do in this same file.
I have also thought we may change the default to audit on.

**Pull request merged**